dask/dataframe/groupby.py

```diff
@@ -293,6 +293,95 @@ def _var_agg(g, levels, ddof):
     return result
 
 
+def _cov_combine(g, levels):
+    return g
+
+
+def _cov_finalizer(df, cols):
+    vals = []
+    num_elements = len(list(it.product(cols, repeat=2)))
+    num_cols = len(cols)
+    vals = list(range(num_elements))
+    col_idx_mapping = dict(zip(cols, range(len(cols))))
+    for i, j in it.combinations_with_replacement(df[cols].columns, 2):
+        x = col_idx_mapping[i]
+        y = col_idx_mapping[j]
+        idx = x + num_cols * y
+        mul_col = "%s%s" % (i, j)
+        ni = df["%s-count" % i]
+        nj = df["%s-count" % j]
+
+        n = np.sqrt(ni * nj)
+        div = n - 1
+        div[div < 0] = 0
+        val = (df[mul_col] - df[i] * df[j] / n).values[0] / div.values[0]
+
+        vals[idx] = val
+        if i != j:
+            idx = num_cols * x + y
+            vals[idx] = val
+
+    level_1 = cols
+    index = pd.MultiIndex.from_product([level_1, level_1])
+    return pd.Series(vals, index=index)
+
+
+def mul_cols(df, cols):
+    _df = type(df)()
+    for i, j in it.combinations_with_replacement(df[cols].columns, 2):
+        col = "%s%s" % (i, j)
```

Comment:
Are we using user-provided columns here? This is an edge case, but you could end up with user columns like `['a', 'b', '']`. Which IIUC, may end up with a name collision, when i or j is the empty string.

Is it possible to just use integers now, and get back the user provided column names just before returning?

Reply:
I am going through this now but it's a little tricky becuase you may have to munge the index as well. For example, we need to handle groupby(['a', 'b']) where a->1, b->2, c->3

Reply:
handled in commit https://github.com/dask/dask/pull/4889/commits/540492849ad26afe87be23a4af6dbd9f64d29f72